### PR TITLE
[MNT] Initializing metrics to "mean_squared_error" In Deep Learning Models

### LINKED ISSUE #1230: [MNT] Cleanup Deep Learning Models
### Describe the issue

In many of our deep learning models, we have the keyword argument `metrics = None`. Then we use the following code: 
```
if self.metrics is None:
            metrics = ["accuracy"]
else:
            metrics = self.metrics
```
Instead we could directly start with `metrics = ["accuracy"]` for brevity and remove the above given block. 

### Suggest a potential alternative/fix

Change as given above!

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by harshithasudhakar**
#### Reference Issues/PRs

Fixes #1230 

#### What does this implement/fix? Explain your changes.

Modified the __init__() such that it sets metrics="mean_squared_error" by default in case of no metric passed. In case of multiple metrics passed, it will be taken as a list of metrics(strings). 

Also removed the code block that checks if metrics is none, and updated the docstring.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

### CODE REVIEW

File: aeon/regression/deep_learning/_encoder.py [hunk 1]
```diff
@@ -117,7 +120,7 @@ def __init__(
         last_file_name="last_model",
         verbose=False,
         loss="mean_squared_error",
-        metrics=None,
+        metrics=["accuracy"],
```
[the remark this answers is not in the document]
Comment (harshithasudhakar) [reply aeon/regression/deep_learning/_encoder.py]: Yes, it is. Thanks for spotting it!

File: aeon/regression/deep_learning/_tapnet.py [hunk 2]
```diff
@@ -101,7 +103,7 @@ def __init__(
         padding="same",
```
Comment (harshithasudhakar) [inline aeon/regression/deep_learning/_tapnet.py:103]: Some of the parameters seem to be missing in the docstring: ‎rp_params, ‎padding‎, ‎callbacks

**Comment by AnonymousCodes911** [comment]
@harshithasudhakar , pre-commit is prompting these:
`aeon/regression/deep_learning/_encoder.py:204:21: F821 undefined name 'metrics' `
`aeon/regression/deep_learning/_fcn.py:204:21: F821 undefined name 'metrics' `
`aeon/regression/deep_learning/_mlp.py:166:21: F821 undefined name 'metrics' `
`aeon/regression/deep_learning/_resnet.py:224:21: F821 undefined name 'metrics' `
`aeon/regression/deep_learning/_tapnet.py:192:21: F821 undefined name 'metrics'` 
but can these be just managed by editing the function?
 `
         model.compile(
            loss=self.loss,
            optimizer=self.optimizer_,
            metrics=metrics,  # should this be metrics="accuracy"?
        )
 `

**Comment by hadifawaz1999** [comment]
keras only accepts a list of metrics, which is why originally the parameter is default to a list. An inner conversion to a list should be done, its an obligation from keras, even if one metric is used

File: aeon/regression/deep_learning/_tapnet.py (hunk 2 above)
Comment (MatthewMiddlehurst) [reply aeon/regression/deep_learning/_tapnet.py:103]: Feel free to fix if they are missing.

**Comment by MatthewMiddlehurst** [comment]
An issue i think you may run into with the current method is the parameter attribute must be unchanged.

So if the input is "accuracy", `self.metrics` must also be "accuracy". You can either create `self._metrics` or do the processing in `_fit`. Feel free to ask questions if anything is confusing.

**Comment by harshithasudhakar** [comment]
> An issue i think you may run into with the current method is the parameter attribute must be unchanged.
> 
> So if the input is "accuracy", `self.metrics` must also be "accuracy". You can either create `self._metrics` or do the processing in `_fit`. Feel free to ask questions if anything is confusing.

Thanks for your suggestion, I'll modify _fit() accordingly

**Comment by harshithasudhakar** [comment]
`_fcn.py` is raising an AssertionError as the metrics attribute is modified in `_fit()`. So, I created a `self._metrics` attribute that holds the list version of the metrics. And made sure `self._metrics` is passed to `build_model`. This should resolve the error as `metrics` is not modified.

**Comment by harshithasudhakar** [comment]
> if a variable is set to private with self, it should not be overridden throughout all the code, for this reason we create self.metrics_ It is still the case in some of the models, once this is fixed it should pass the tests

Yes, I have fixed it all now :)

**Comment by harshithasudhakar** [comment]
@MatthewMiddlehurst @hadifawaz1999 @AnonymousCodes911 would be nice if you can review this pr 😁

**Comment by AnonymousCodes911** [comment]
@harshithasudhakar , 
I am just a fellow contributor 😅  , @hadifawaz1999 and @MatthewMiddlehurst are the maintainers!

File: aeon/regression/deep_learning/_encoder.py [hunk 3]
```diff
@@ -70,8 +70,11 @@ class EncoderRegressor(BaseDeepRegressor):
         Seed for random number generation.
     loss:
         The loss function to use for training.
-    metrics:
-        The evaluation metrics to use during training.
+    metrics: str or list of str, default=["accuracy"]
+        The evaluation metrics to use during training. If
+        a single string metric is provided, it will be
+        used as the only metric. If a list of metrics are
+        provided, all will be used for evaluation.
```
Comment (MatthewMiddlehurst) [inline aeon/regression/deep_learning/_encoder.py:77]: The default should be changed here. Make sure you update the documentation for other DL estimators which have been changed also.

File: aeon/regression/deep_learning/_encoder.py [hunk 4]
```diff
@@ -202,7 +200,7 @@ def build_model(self, input_shape, **kwargs):
         model.compile(
             loss=self.loss,
             optimizer=self.optimizer_,
-            metrics=metrics,
+            metrics=self.metrics,
```
Comment (MatthewMiddlehurst) [inline aeon/regression/deep_learning/_encoder.py:203]: This one has not been updated, the changes for each estimator should be different.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
The currently changes estimators are regressors, is accuracy the correct metric to use @hadifawaz1999?

**Comment by harshithasudhakar** [comment]
@MatthewMiddlehurst Hi. I am new to DL, but I'm eager to dive deep. To tailor each of these models according to it's specific characteristics, I would require to read more and understand the models better. Are there any supporting docs or links I can refer to, to contribute effectively? Would greatly appreciate if the community can help me with this :)

**Comment by hadifawaz1999** [comment]
> The currently changes estimators are regressors, is accuracy the correct metric to use @hadifawaz1999?

No it should be `mean_squared_error` instead

**Comment by hadifawaz1999** [comment]
> @MatthewMiddlehurst Hi. I am new to DL, but I'm eager to dive deep. To tailor each of these models according to it's specific characteristics, I would require to read more and understand the models better. Are there any supporting docs or links I can refer to, to contribute effectively? Would greatly appreciate if the community can help me with this :)

Hello @harshithasudhakar yes you can refer to the deep learning for time series classification review in 2019. [link](https://arxiv.org/abs/1809.04356)

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
In the case of regression, the default metric should be "mean_squared_error"

**Comment by harshithasudhakar** [comment]
> ###
> [Mar 25, 2024](#issuecomment-2017314004)

Thanks! I'll look into it!

**Review by MatthewMiddlehurst — COMMENTED** [review]
Two of the docstrings of look like they still need to be updated also for metrics. Think this is close though, thanks for the work.

**Comment by harshithasudhakar** [comment]
> It looks like the old variable for setting the metric is stiff there, it should be removed.
> 
> Please remove the bits of code added setting random states etc., I think there were probably accidentally added during a merge.

Yes, there were accidental changes but I fixed them now.

**Review by MatthewMiddlehurst — APPROVED** [review]
LGTM, thanks.

File: aeon/regression/deep_learning/_cnn.py [hunk 5]
```diff
@@ -68,7 +68,8 @@ class CNNRegressor(BaseDeepRegressor):
     loss            : string, default="mean_squared_error"
         fit parameter for the keras model
     optimizer       : keras.optimizer, default=keras.optimizers.Adam(),
-    metrics         : list of strings, default=["accuracy"],
+    metrics         : list of strings, default="mean_squared_error",
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_cnn.py:71]: should add here "If
        a single string metric is provided, it will be
        used as the only metric. If a list of metrics are
        provided, all will be used for evaluation."

File: aeon/regression/deep_learning/_tapnet.py [hunk 6]
```diff
@@ -46,8 +48,12 @@ class TapNetRegressor(BaseDeepRegressor):
         activation function for the last output layer
     loss                : str, default = "mean_squared_error"
         loss function for the classifier
+    metrics             : list of str, default = "mean_squared_error"
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_tapnet.py:51]: should add here "If
        a single string metric is provided, it will be
        used as the only metric. If a list of metrics are
        provided, all will be used for evaluation."

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
Thanks, just a couple last changes

**Review by hadifawaz1999 — APPROVED** [review]
lgtm thanks

**Comment by MatthewMiddlehurst** [comment]
@all-contributors add @harshithasudhakar for code
